h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoost.java

```diff
@@ -438,13 +442,94 @@ private void scoreAndBuildTrees(final XGBoostSetupTask setupTask, final BoosterP
           break;
         }
       }
+
+      Map<String, Integer> monotoneConstraints = _parms.monotoneConstraints();
+      if ((!monotoneConstraints.isEmpty()) && constraintCheckEnabled()) {
+        checkConstraints(model.model_info(), monotoneConstraints);
+      }
+      
       _job.update(0, "Scoring the final model");
       // Final scoring
       doScoring(model, boosterProvider, true);
       // Finish remaining work (if stopped early)
       _job.update(_parms._ntrees-model._output._ntrees);
     }
 
+    private boolean constraintCheckEnabled() {
+      return Boolean.parseBoolean(getSysProperty("xgboost.monotonicity.checkEnabled", "true"));
+    }
+
+    private void checkConstraints(XGBoostModelInfo model_info, Map<String, Integer> monotoneConstraints) {
+      GradBooster booster = XGBoostJavaMojoModel.makePredictor(model_info._boosterBytes).getBooster();
+      if (!(booster instanceof GBTree)) {
```

Comment:
This is not ideal because the check can stop working over the time, instead
- we should not invoke monotonicity check for `gblinear`,
- change this into IllegalStateException.

Reply:
fixed